projects/igniteui-angular/src/lib/core/selection.ts

```diff
@@ -1,141 +1,98 @@
         this.selection.set(componentID, currSelection);
     }
 
     public get_selection_length(componentID: string): number {
-        return (this.get_selection(componentID) || []).length;
+        const sel = this.get_selection(componentID);
+        return sel ? sel.size : 0;
     }
 
     /**
      * Add item to component selection using item's ID.
      * @param componentID ID of the component, which we add new item to.
      * @param itemID ID of the item to add to component selection.
-     * @param compareFunc Custom function that search for already existing item with the provided ID.
-     *     If not presented then item is compared by reference (using ===).
-     * @param currSelection Used internally only by the selection (select_items method) to accumulate selection for multiple items.
+     * @param sel Used internally only by the selection (select_items method) to accumulate selection for multiple items.
      *
      * @returns selection after the new item is added
      */
-    public select_item(componentID: string, itemID, compareFunc?: (item) => boolean, currSelection?: any[]): any[] {
-        let itemPresented;
-        if (!currSelection) {
-            currSelection = this.get_selection(componentID);
-        }
-        if (currSelection === undefined) {
-            currSelection = [];
+    public select_item(componentID: string, itemID, sel?: Set<any>, clearSelection?: boolean): Set<any> {
+        if (!sel) {
+            sel = this.get_selection(componentID);
         }
-        currSelection = [...currSelection];
-
-        if (!compareFunc) {
-            itemPresented = currSelection.indexOf(itemID) !== -1;
-        } else {
-            itemPresented = currSelection.find(compareFunc) !== undefined;
+        if (sel === undefined || clearSelection) {
+            sel = new Set();
         }
-        if (!itemPresented) {
-            currSelection.push(itemID);
-        }
-        return currSelection;
+        sel.add(itemID);
+        return sel;
     }
 
     /**
      * Add items to component selection using their IDs.
      * @param componentID ID of the component, which we add new items to.
      * @param itemIDs Array of IDs of the items to add to component selection.
-     * @param compareFunc Custom function that search for already existing items with the provided IDs.
-     * If not presented then items are compared by reference (using ===).
      *
      * @returns selection after the new item is added
      */
-    public select_items(componentID: string, itemIDs: any[], compareFunc?: (item) => boolean): any[] {
-        let selection: any[];
-        itemIDs.forEach((item) => selection = this.select_item(componentID, item, compareFunc, selection));
+    public select_items(componentID: string, itemIDs: any[], clearSelection?: boolean): Set<any> {
+        let selection: Set<any>;
+        itemIDs.forEach((item) => selection = this.select_item(componentID, item, selection, clearSelection));
```

Comment:
I think that calling `select_item` with `clearSelection === true` will always add clear selection to an empty Set, so iwhen calling `select_items('grid1', ['1','2','3'], true)`, the selection in the end will be `Set('3')`